docarray/array/mixins/setitem.py

```diff
@@ -76,127 +76,16 @@ def __setitem__(
-                        )
-                elif isinstance(index[0], (slice, Sequence)):
-                    _attrs = index[1]
-
-                    if isinstance(_attrs, str):
-                        # a -> [a]
-                        # [a, a] -> [a, a]
-                        _attrs = (index[1],)
-                    if isinstance(value, (list, tuple)) and not any(
-                        isinstance(el, (tuple, list)) for el in value
-                    ):
-                        # [x] -> [[x]]
-                        # [[x], [y]] -> [[x], [y]]
-                        value = (value,)
-                    if not isinstance(value, (list, tuple)):
-                        # x -> [x]
-                        value = (value,)
-
-                    _docs = self[index[0]]
-                    for _a, _v in zip(_attrs, value):
-                        if _a in ('tensor', 'embedding'):
-                            if _a == 'tensor':
-                                _docs.tensors = _v
-                            elif _a == 'embedding':
-                                _docs.embeddings = _v
-                            for _d in _docs:
-                                self._set_doc_by_id(_d.id, _d)
-                        else:
-                            if len(_docs) == 1:
-                                self._set_doc_attr_by_id(_docs[0].id, _a, _v)
-                            else:
-                                for _d, _vv in zip(_docs, _v):
-                                    self._set_doc_attr_by_id(_d.id, _a, _vv)
+            if isinstance(index, tuple) and len(index) == 2:
+                self._set_by_pair(index[0], index[1], value)
+
             elif isinstance(index[0], bool):
-                if len(index) != len(self):
-                    raise IndexError(
-                        f'Boolean mask index is required to have the same length as {len(self._data)}, '
-                        f'but receiving {len(index)}'
-                    )
-                _selected = itertools.compress(self, index)
-                self._set_doc_value_pairs(_selected, value)
+                self._set_by_mask(index, value)
+
             elif isinstance(index[0], (int, str)):
-                if not isinstance(value, Sequence) or len(index) != len(value):
-                    raise ValueError(
-                        f'Number of elements for assigning must be '
-                        f'the same as the index length: {len(index)}'
-                    )
-                if isinstance(value, Document):
-                    for si in index:
-                        self[si] = value  # leverage existing setter
-                else:
-                    for si, _val in zip(index, value):
-                        self[si] = _val  # leverage existing setter
+                for si, _val in zip(index, value):
+                    self[si] = _val  # leverage existing setter
```

Comment:
Invalid types in the sequence should be caught and not fail silently; particularly things like da[1.0, 2.0] coming from np or torch might be a pitfall.

```suggestion
            else:
                raise IndexError(f"{index} should be either a sequence of bool, int or str")
```

Reply:
oups not used to it yet :confused:

Reply:
fixed